Introduce feature flags

senadir opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->
This PR introduces a simple feature flag and gate features, as a start, it will only prevent Checkout and Cart from being registered.

It introduces ~~3 flags~~ 2 flags: ~~Core,~~ Stable, and Experimental.

Any code not gated is considered stable enough to be included in Woo Core.

Stable code is code that is not so stable to be included in Core, but stable enough to be shipped to the plugin, in the future, an option could be added to Woo Settings to enable this code.

Experimental is code that is not stable yet to be shipped, think Cart and Checkout blocks by the end of last cycle.

For development purposes, the default flag will be `experimental`.

We also append `WOOCOMMERCE_BLOCKS_PHASE="stable"` to the build command in `npm run deploy`.

To achieve this, I used a combo of `DefinePlugin` of webpack and an env variable `WOOCOMMERCE_BLOCKS_PHASE`.


This code heavily inspires from https://github.com/WordPress/gutenberg/pull/13324
<!-- Reference any related issues or PRs here -->
Fixes #1500

senadir commented:
This is set to `Draft` and `in Progress` because I didn’t yet manage to get it to work, I only tagged for review to help determine why it’s not working (something specific about our build process?)

senadir commented:
Moving the config to `webpack-helper` fixed the issue, all working right now.

nerrad commented on package.json:
```diff
@@ -19,9 +19,12 @@
 	},
+	"config": {
+		"WOOCOMMERCE_BLOCKS_PHASE": "experimental"
+	},
```
This works for Gutenberg because the [WordPress core package.json](https://github.com/WordPress/wordpress-develop/blob/master/package.json) doesn't define this configuration so the default in the Gutenberg config falls back to `1`.  The blocks feature plugin is included in WooCommerce core differently via composer pulling in a tagged release that is _already built_.  So effectively what you've implemented here means that `experimental` is _always_ used (I verified by doing `cross env WOOCOMMERCE_BLOCKS_PHASE=stable npm run build`).

Instead, what you want to do is remove this configuration and _only_ configure the feature gating via the environment variable.  The `DefinePlugin` definition would have to be something like this instead:

```js
new DefinePlugin( {
				// Inject the `WOOCOMMERCE_BLOCKS_PHASE` global, used for feature flagging.
				'process.env.WOOCOMMERCE_BLOCKS_PHASE': JSON.stringify(
					process.env.WOOCOMMERCE_BLOCKS_PHASE ||
						'experimental'
				),
			} ),
```

senadir commented:
So my latest changes are not complete.

I addressed the feedback and added even more gating, since we output a js file for each block (well, 2 files), i moved those to `experimentalFrontendEntry` and `experimentalMainEntry`, I left the stable ones in `stableFrontendEntry` and `stableMainEntry`, and according to the flag, I merge them.

this will cause some html errors since PHP will enqueue a nonexistent file, I'm adding some feature gating to PHP as well, but we still need a way to define the constant at run time.

senadir commented on src/RestApi.php:
```diff
@@ -84,24 +84,33 @@ public static function maybe_init_cart_session( $return, $request = false ) {
+		$experimental_controllers = [
+			'store-cart'       => __NAMESPACE__ . '\RestApi\StoreApi\Controllers\Cart',
+			'store-cart-order' => __NAMESPACE__ . '\RestApi\StoreApi\Controllers\CartOrder',
+		];
```
@mikejolley those controllers only affect the cart and checkout endpoints right?

mikejolley replied:
They are built for that purpose yes. Should ensure these files are included during unit tests though.

nerrad replied:
I don't think it's a big deal if the REST endpoints are gated or not as they are explicitly listed as being private anyways.  So I would recommend just not gating the endpoints. If you decide to gate them, then you'll need to gate `store-cart-items`, `store-cart-coupons`, `store-cart-shipping-rates` and `store-customer` too.

senadir commented on woocommerce-gutenberg-products-block.php:
```diff
@@ -18,6 +18,8 @@
 defined( 'ABSPATH' ) || exit;
 
+define( 'WOOCOMMERCE_BLOCKS_PHASE', 'experimental' );
+
```
Currently, the way we define the flag in PHP is hacky, I couldn’t find a way yet to do this via Composer or something, so in the future when you want to build a stable release you need to manually change this.
I'm thinking we use something like the one that changes version numbers. @woocommerce/blocks

mikejolley replied:
What about enabling this based on the existence of a built file? To avoid any manual changes.

nerrad replied:
💯 to Mike's reply here! You could maybe use this webpack plugin: https://www.npmjs.com/package/create-file-webpack and create a php file at the end of the build process that basically defines the `WOOCOMMERCE_BLOCKS_PHASE` constant and then you just load that file in the bootstrap.  Note, you should load this as a part of the `src/Domain/Bootstrap` class because the main plugin file (`woocommerce-gutenberg-products-block`) is not loaded when in WooCommerce core.

nerrad commented on bin/webpack-helpers.js:
```diff
@@ -127,22 +127,36 @@ const mainEntry = {
+const mainEntry =
+	process.env.WOOCOMMERCE_BLOCKS_PHASE === 'experimental'
+		? Object.assign( {}, stableMainEntry, experimentalMainEntry )
+		: Object.assign( {}, stableMainEntry );
```
You can just use object spread here if you want. But this also is okay.

senadir replied:
forget that was a thing 🤦🏻‍♂️

nerrad commented on bin/webpack-helpers.js:
```diff
@@ -127,22 +127,36 @@ const mainEntry = {
+};
 
-	// cart & checkout blocks
+const experimentalMainEntry = {
```
Great idea to split out the stable from experimental builds!

nerrad commented on package.json:
```diff
@@ -22,11 +22,11 @@
 		"build": "cross-env BABEL_ENV=default NODE_ENV=production webpack",
 		"build:ci": "npm run build && npm run size-check && npm run lint:js && npm run lint:css",
-		"start": "cross-env BABEL_ENV=default webpack --watch --info-verbosity none",
+		"start": "cross-env WOOCOMMERCE_BLOCKS_PHASE=experimental BABEL_ENV=default webpack --watch --info-verbosity none",
```
Is this needed if the default is `experimental` (`npm build` or `npm build:ci` executed on their own will be using `experimental` too)?

senadir replied:
now that we have default values, we don’t need to pass this yeah.

nerrad commented on bin/webpack-helpers.js:
```diff
@@ -136,8 +136,8 @@ const experimentalMainEntry = {
-		? Object.assign( {}, stableMainEntry, experimentalMainEntry )
-		: Object.assign( {}, stableMainEntry );
+		? { ...stableMainEntry, ...experimentalMainEntry }
+		: { ...stableMainEntry };
```
Does the last condition need to be a new object? Probably can just be `stableMainEntry`? The same comment applies to the `stableFrontendEntry` implementation.

senadir replied:
get so used to writing immutable objects  😅fixed.

nerrad commented on src/Domain/Bootstrap.php:
```diff
@@ -148,4 +150,12 @@ protected function remove_core_blocks() {
+	protected function define_feature_flag() {
+		$woo_options = parse_ini_file( __DIR__ . '/../../blocks.ini' );
+		define( 'WOOCOMMERCE_BLOCKS_PHASE', $woo_options['woocommerce_blocks_phase'] );
+	}
```
Should validate `$woo_options` isn't `false` and is an array before using it.

nerrad commented on blocks.ini:
```diff
@@ -0,0 +1 @@
+woocommerce_blocks_phase = experimental
```
I think this file should be added to `.gitignore` and not included in repo. It's more of a build artifact. This of course _does_ mean you need to address my comment with the `Bootstrap.php` changes for validating the value from `blocks.ini` before using it.

nerrad commented on webpack.config.js:
```diff
@@ -90,6 +91,14 @@ const CoreConfig = {
+			// content of the file
+			content: `woocommerce_blocks_phase = ${ process.env
+				.WOOCOMMERCE_BLOCKS_PHASE || 'experimental' }`,
+		} ),
```
Nice!

nerrad commented on src/Domain/Bootstrap.php:
```diff
@@ -148,4 +150,12 @@ protected function remove_core_blocks() {
+	 */
+	protected function define_feature_flag() {
+		$woo_options = parse_ini_file( __DIR__ . '/../../blocks.ini' );
+		define( 'WOOCOMMERCE_BLOCKS_PHASE', $woo_options['woocommerce_blocks_phase'] );
```
We also should make sure that `$woo_options['woocommerce_blocks_phase']` is a string and maybe sanitized? Should it be checked against a whitelisted set of values?

senadir commented:
@nerrad addressed the latest comments, this is ready for another spin

nerrad commented on src/Domain/Bootstrap.php:
```diff
@@ -156,6 +156,13 @@ protected function remove_core_blocks() {
+		) ) {
+			$flag = sanitize_key( $woo_options['woocommerce_blocks_phase'] );
+		}
+		define( 'WOOCOMMERCE_BLOCKS_PHASE', $flag );
```
If `$flag` defaults to stable, then you only have the check for `experimental` to override right?  Also, no need to run though `sanitize_key` if you are doing a whitelist.  So you could probably simplify this by doing:

```php
$flag = is_array( $woo_options ) && 'experimental' === $woo_options['woocommerce_blocks_phase'] ? 'experimental' : 'stable';
define( 'WOOCOMMERCE_BLOCKS_PHASE', $flag );
```

senadir replied:
I applied your fix and added the env to travis, I did a check first for the env variable than to the blocks.ini, it should be safe for any environment.

```php
if ( getenv( 'WOOCOMMERCE_BLOCKS_PHASE' ) ) {
	$flag = getenv( 'WOOCOMMERCE_BLOCKS_PHASE' );
} else {
	$woo_options = parse_ini_file( __DIR__ . '/../../blocks.ini' );
	$flag        = is_array( $woo_options ) && 'experimental' === $woo_options['woocommerce_blocks_phase'] ? 'experimental' : 'stable';
}
```

nerrad commented on src/Domain/Bootstrap.php:
```diff
@@ -155,13 +155,11 @@ protected function remove_core_blocks() {
-		) ) {
-			$flag = sanitize_key( $woo_options['woocommerce_blocks_phase'] );
+		if ( getenv( 'WOOCOMMERCE_BLOCKS_PHASE' ) ) {
+			$flag = getenv( 'WOOCOMMERCE_BLOCKS_PHASE' );
```
So the only problem here, is while `$woo_options` is sanitized (only allows `experimental` as the value), you're assigning the environment variable for the flag without any sanitization. I'd suggest something like:

```php
$allowed_flags = [ 'experimental', 'stable' ];
$flag = getenv( 'WOOCOMMERCE_BLOCKS_PHASE' );
if ( ! in_array( $flag, $allowed_flags ) ) {
    $woo_options = parse_ini_file( __DIR__ . '/../../blocks.ini' );
			$flag        = is_array( $woo_options ) && 'experimental' === $woo_options['woocommerce_blocks_phase'] ? 'experimental' : 'stable';
}
define( 'WOOCOMMERCE_BLOCKS_PHASE', $flag );
```

senadir replied:
Interesting, I assumed env must be safe but we can’t know for sure, I'm going to roll with your suggestion here.

nerrad replied:
One could definitely argue that if someone has gained access to a system and is able to set environment variables, you have a bigger problem. I generally just like to be in the habit of sanitizing _any_ user space type data. We know exactly what data we want here, so let's make sure it's valid. Our future selves might thank us.
